lib/collection/src/lookup.rs

```diff
@@ -0,0 +1,76 @@
+use std::collections::HashMap;
+
+use futures::Future;
+use itertools::Itertools;
+use schemars::JsonSchema;
+use segment::data_types::groups::PseudoId;
+use segment::types::{PointIdType, WithPayloadInterface, WithVector};
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLockReadGuard;
+
+use crate::collection::Collection;
+use crate::operations::consistency_params::ReadConsistency;
+use crate::operations::types::{CollectionError, CollectionResult, PointRequest, Record};
+use crate::shards::shard::ShardId;
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(untagged)]
+pub enum Lookup {
+    None,
+    Single(Record),
+}
+
+impl From<Record> for Lookup {
+    fn from(record: Record) -> Self {
+        Lookup::Single(record)
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct LookupRequest {
+    pub collection_name: String,
+    pub with_payload: WithPayloadInterface,
+    pub with_vectors: WithVector,
+}
+
+pub async fn lookup_ids<'a, F, Fut>(
+    request: LookupRequest,
+    values: Vec<PseudoId>,
+    collection_by_name: F,
+    read_consistency: Option<ReadConsistency>,
+    shard_selection: Option<ShardId>,
+) -> CollectionResult<HashMap<PseudoId, Lookup>>
+where
+    F: FnOnce(String) -> Fut,
+    Fut: Future<Output = Option<RwLockReadGuard<'a, Collection>>>,
+{
+    let collection = collection_by_name(request.collection_name.clone())
```

Comment:
Hmm! Good point, I like that

Reply:
Though this would maybe require to acquire the lock of the lookup collection since the start of the grouping.

I think a good way to go is to keep this as-is, and include validation only earlier in the chain (without acquiring the lock).